### PR TITLE
Remove `try-except` from gridsearch 

### LINKED ISSUE #157: Make try-except statement in `_gridsearch` more specific
Currently, if there is, e.g., any `RuntimeError` in the `loss`, `_gridsearch` will silently fail and just return the boundary value of the grid; see [here](https://github.com/aleximmer/Laplace/blob/main/laplace/baselaplace.py#L288-L302). One example for this is when the loss function is not appropriate for the setting.

### PR DESCRIPTION
**Pull request by wiseodd**
Fixes #157 and #194 

I added new test cases and also made sure to use the classification likelihood during gridsearch in the case of reward modeling.

### CODE REVIEW

File: laplace/baselaplace.py [hunk 1]
```diff
@@ -429,21 +438,20 @@ def _gridsearch(
         results = list()
         prior_precs = list()
         pbar = tqdm.tqdm(interval) if progress_bar else interval
+
         for prior_prec in pbar:
             self.prior_precision = prior_prec
-            try:
```
Comment (runame) [inline laplace/baselaplace.py:434]: Maybe it's worth keeping the try-except statement for its original purpose, to catch inversion errors for a prior precision that is too small, so that the gridsearch is not interrupted, but the next value is tried. So it could be something like this (this would work when using Cholesky for the inversions):
```python
try:
    ...
except RuntimeError as error:
    if "not positive-definite" in str(error):
        result = np.inf
    else:
        raise error
```

**Review by runame — COMMENTED** [review]
See #200 for a suggested refactor and my comment below.

File: laplace/baselaplace.py [hunk 2]
```diff
@@ -442,8 +454,13 @@ def _gridsearch(
                     loss_with_var=loss_with_var,
                     dict_key_y=self.dict_key_y,
                 )
-            except RuntimeError:
+            except LinAlgError:
```
Comment (wiseodd) [inline laplace/baselaplace.py:457]: @runame please recheck. It seems PyTorch now uses `LinAlgError` for this https://github.com/pytorch/pytorch/pull/68571. But I also added your suggestion for safety.

**Review by runame — APPROVED** [review]
LGTM now!
